### PR TITLE
Initial TypeScript typings

### LINKED ISSUE #115: Module missing a declaration - AniLink
Not sure what the issue is- I have a fairly bare bones gatsby ecommerce starter kit set up. I tried to use the AniLink transition to add some spice but I get this error whenever I try and import the model after installing transition link and gsap. 

```
Could not find a declaration file for module 'gatsby-plugin-transition-link/AniLink'. '/Users/robert/Software/bDev/evu/node_modules/gatsby-plugin-transition-link/AniLink/index.js' implicitly has an 'any' type.
  Try `npm install @types/gatsby-plugin-transition-link` if it exists or add a new declaration (.d.ts) file containing `declare module 'gatsby-plugin-transition-link/AniLink';`ts(7016)
```

the import line:

`import AniLink from 'gatsby-plugin-transition-link/AniLink';`

Usage: 

`<AniLink paintDrip to="/allProducts"><button class="buttonMain">View Now</button></AniLink>`

package.json dependencies: 

```
  "gsap": "^2.1.3",
  "gatsby-plugin-transition-link": "^1.12.4",
```

Config File:

```
require("dotenv").config()


module.exports = {
  siteMetadata: {
    title: `EssentialVu`,
    description: `Some really sweet ecommerce site.`,
    author: `@gatsbyjs`,
  },
  
  plugins: [
    `gatsby-plugin-react-helmet`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `images`,
        path: `${__dirname}/src/images`,
      },
    },
    `gatsby-transformer-sharp`,
    `gatsby-plugin-stripe`,
    {
      resolve: `gatsby-source-stripe`,
      options: {
        objects: ["Sku"],
        secretKey: process.env.STRIPE_SECRET_KEY,
        downloadFiles: true,
      },
    },
    {
      resolve: "gatsby-plugin-transition-link",
      options: {
          layout: require.resolve(`./src/components/layout.js`)
        },
    }
    `gatsby-plugin-sharp`,
    {
      resolve: `gatsby-plugin-manifest`,
      options: {
        name: `gatsby-starter-default`,
        short_name: `starter`,
        start_url: `/`,
        background_color: `#663399`,
        theme_color: `#663399`,
        display: `minimal-ui`,
        icon: `src/images/gatsby-icon.png`, // This path is relative to the root of the site.
      },
    },
    // this (optional) plugin enables Progressive Web App + Offline functionality
    // To learn more, visit: https://gatsby.dev/offline
    // `gatsby-plugin-offline`,
  ],
  
}
```

### PR DESCRIPTION
**Pull request by crutchcorn**
Like most software, this can be improved in a few ways and likely includes some form of errors. Review from someone more familiar with the codebase would be appreciated.

I know one of the ways these typings can be made to be more strict would be to decouple state of entry and exit and make them persist as their own meta types, which would enforce your inability to access state that's only present on `entry` from a `trigger` function accessing an `exit` state, but that's a level of complexity in typings that I wouldn't be comfortable introducing as an initial set of typings

Closes #115 #155

### CODE REVIEW

**Comment by TylerBarnes** [comment]
Hi @crutchcorn , sorry for the extremely delayed response.
To this day I haven't done much work with Typescript. Are you able to provide me some simple steps to get started trying out these typings in a project?
I'm the only one deeply familiar with the codebase as far as I'm aware but like I mentioned I've only used Typescript a handful of times.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
+/**
+ * This should be removed once added upstream:
+ * https://github.com/TylerBarnes/gatsby-plugin-transition-link/pull/202
+ */
+
+declare module 'gatsby-plygin-transition-link/AniLink' {
```
Comment (dvzrd) [inline index.d.ts:6]:
```suggestion
declare module 'gatsby-plugin-transition-link/AniLink' {
```

Comment (dvzrd) [reply index.d.ts:6]: found a typo 8)

**Review by dvzrd — COMMENTED** [review]
Is this missing anything? Haven't really tested it.

**Review by felipe-heredia — APPROVED** [review]

**Comment by felipesuri** [comment]
@TylerBarnes I'ts workind, I tested it on my project.

**Comment by felipesuri** [comment]
@crutchcorn add this lines on the AniLinkProps interface:

```ts
title?: string;
target?: string;
```

**Comment by chrisbward** [comment]
any joy?

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		bg?: string;
+	}
+	type AniLinkTypes = AniLinkFade | AniLinkPaintDrip | AniLinkSwipe | AniLinkCover;
+
+	interface AniLinkProps {
+		to: string;
+		direction?: string;
+		duration?: number;
+		top?: 'exit' | 'entry';
+		entryOffset?: number;
```
Comment (crutchcorn) [inline index.d.ts:36]:
```suggestion
		entryOffset?: number;
		title?: string;
		target?: string;
```

**Comment by crutchcorn** [comment]
@TylerBarnes I am so sorry for not responding when you'd made a comment. I spaced on it as I had a lot going on at the time and it just fell from my radar.

The best way to test this would be to either:

1) Load up a TypeScript project that uses `gatsby-plugin-transition-link` and see if the code works as-intended
2) Add TypeScript type tests using [`dtslint`](https://github.com/Microsoft/dtslint). An example of that can be seen here:

https://github.com/plopjs/node-plop/tree/master/types

I can setup `dtslint` tests if that would be helpful. I admittedly didn't know how to do that at the initial time of this PR

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		title?: string;
+		target?: string;
+	}
+	class AniLink extends Component<AniLinkProps & AniLinkTypes> {}
```
Comment (alexcmgit) [inline index.d.ts:40]: In my project I was trying to use as:

```tsx
import React from "react";
import AniLink from "gatsby-plugin-transition-link/AniLink";

export type IInLink = {
  to: string;
};

export function InLink({ to, children }: React.PropsWithChildren<IInLink>) {
  return (
    <AniLink fade to={to}>
      {children}
    </AniLink>
  );
}
```

Got:

```txt
Type '{ children: ReactNode; fade: true; to: string; }' is not assignable to type 'IntrinsicAttributes & (IntrinsicClassAttributes<AniLink> & Readonly<AniLinkProps & AniLinkTypes>)'.
      Property 'children' does not exist on type 'IntrinsicAttributes & (IntrinsicClassAttributes<AniLink> & Readonly<AniLinkProps & AniLinkTypes>)'.
```

Solved by adding children to the declaration:

```suggestion
	class AniLink extends Component<
  React.PropsWithChildren<AniLinkProps> & AniLinkTypes
  > {}
```

**Pull request closed without merging**